templates/cinder/bin/healthcheck.py

```diff
@@ -0,0 +1,162 @@
+        # of dict/filters, and for others is a single element with the filters.
+        cls.services_filters = services_filters
+
+    @staticmethod
+    def check_service(services, **filters):
+        # Check services DB connectivity by looking at their DB heartbeat
+        our_services = [service for service in services
+                        if all(getattr(service, key, None) == value
+                               for key, value in filters.items())]
+        if not our_services:
+            raise Exception('Not found', f'Service not found with: {filters}')
+
+        num_services = len(our_services)
+        if num_services != 1:
+            print(f'Too many services found! {num_services}')
+
+        ok = all(service.is_up for service in our_services)
+        if not ok:
+            raise Exception('Service error', 'Service is not UP')
+
+        # TODO: Add a RabbitMQ connectivity check, for example using the
+        #       logging get feature
+
+    def do_GET(self):
+        try:
+            services = objects.ServiceList.get_all_by_binary(self.ctxt,
+                                                             self.binary)
+        except Exception as exc:
+            return self.send_error(500,
+                                   'DB access error',
+                                   f'Failed to connect to the database: {exc}')
+
+        for service_filters in self.services_filters:
+            try:
+                self.check_service(services, **service_filters)
+            except Exception as exc:
+                return self.send_error(500, exc.args[0], exc.args[1])
+
+        self.send_response(200)
+        self.send_header("Content-type", "text/html")
+        self.end_headers()
+        self.wfile.write('<html><body>OK</body></html>'.encode('utf-8'))
+
+
+if __name__ == "__main__":
+    if 3 < len(sys.argv) < 2 or sys.argv[1] not in BINARIES:
+        print('Healthcheck requires the binary type as argument (one of: '
+              + ', '.join(BINARIES) +
+              ') and optionally the location of the config file.')
+        sys.exit(1)
+    binary = sys.argv[1]
+
+    cfg_file = sys.argv[2] if len(sys.argv) == 3 else '/etc/cinder/cinder.conf'
+
+    objects.register_all()
+
+    # Due to our imports cinder.common.config.global_opts are registered in
+    # cinder/volume/__init__.py bringing in the following config options we
+    # want: host, storage_availability_zone, service_down_time, cluster,
+    # enabled_backends
+    # Initialize Oslo Config
+    CONF.register_opt(cfg.StrOpt('cluster', default=None))
+    CONF(['--config-file', cfg_file], project='cinder')
+
+    HeartbeatServer.initialize_class(binary)
+
+    webServer = server.HTTPServer((HOSTNAME, SERVER_PORT), HeartbeatServer)
+    print(f"Cinder Healthcheck Server started http://{HOSTNAME}:{SERVER_PORT}")
+
+    try:
+        webServer.serve_forever()
+    except KeyboardInterrupt:
```

Comment:
Might be worth adding a SIGTERM handler. IIRC KeyboardInterrupt is only raised for SIGINT. Container stop will send a SIGTERM and (if still running after a timeout) a SIGKILL.

Reply:
done